src/setup/sound.c

```diff
@@ -169,74 +168,43 @@ static void UpdateMidiDevice(TXT_UNCAST_ARG(widget), TXT_UNCAST_ARG(data))
 static txt_dropdown_list_t *MidiDeviceSelector(void)
 {
     txt_dropdown_list_t *result;
+    int num_devices;
     int all_devices;
-    int device_ids[MAX_MIDI_DEVICES];
-    MMRESULT mmr;
-    MIDIOUTCAPS mcaps;
     int i;
 
-    if (midi_num_devices > 0)
-    {
-        for (i = 0; i < midi_num_devices; ++i)
-        {
-            free(midi_names[i]);
-            midi_names[i] = NULL;
-        }
-        free(midi_names);
-        midi_names = NULL;
-    }
-    midi_num_devices = 0;
+    midi_index = 0;
+    num_devices = 1; // Always show MIDI_MAPPER.
+    all_devices = midiOutGetNumDevs(); // Does not include MIDI_MAPPER.
 
-    // get the number of midi devices on this system
-    all_devices = midiOutGetNumDevs() + 1; // include MIDI_MAPPER
-    if (all_devices > MAX_MIDI_DEVICES)
+    if (all_devices > MAX_MIDI_DEVICES - num_devices)
     {
-        all_devices = MAX_MIDI_DEVICES;
+        all_devices = MAX_MIDI_DEVICES - num_devices;
     }
 
-    // get the valid device ids only, starting from -1 (MIDI_MAPPER)
-    for (i = 0; i < all_devices; ++i)
+    for (i = 0; i < all_devices; i++)
     {
-        mmr = midiOutGetDevCaps(i - 1, &mcaps, sizeof(mcaps));
-        if (mmr == MMSYSERR_NOERROR)
-        {
-            device_ids[midi_num_devices] = i - 1;
-            midi_num_devices++;
-        }
-    }
+        MIDIOUTCAPS caps;
 
-    // get the device names
-    midi_names = malloc(midi_num_devices * sizeof(char *));
-    for (i = 0; i < midi_num_devices; ++i)
-    {
-        mmr = midiOutGetDevCaps(device_ids[i], &mcaps, sizeof(mcaps));
-        if (mmr == MMSYSERR_NOERROR)
+        if (midiOutGetDevCaps(i, &caps, sizeof(caps)) == MMSYSERR_NOERROR)
         {
-            midi_names[i] = M_StringDuplicate(mcaps.szPname);
-        }
-    }
+            free(midi_names[num_devices]);
```

Comment:
What prevents trying to `free` the constant string stored in `midi_names` on initialization? This is not a problem if you remove that from there.